unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/QuestsPanelHUD/QuestsPanelHUDView.cs

```diff
@@ -0,0 +1,135 @@
+
+        internal static QuestsPanelHUDView Create()
+        {
+            var view = Instantiate(Resources.Load<GameObject>(VIEW_PATH)).GetComponent<QuestsPanelHUDView>();
+#if UNITY_EDITOR
+            view.gameObject.name = "_QuestsPanelHUDView";
+#endif
+            return view;
+        }
+
+        public void Awake()
+        {
+            questPopup.gameObject.SetActive(false);
+        }
+
+        public void RequestAddOrUpdateQuest(string questId)
+        {
+            if (questsToBeAdded.Contains(questId))
+                return;
+
+            questsToBeAdded.Add(questId);
+        }
+
+        internal void AddOrUpdateQuest(string questId)
+        {
+            if (!quests.TryGetValue(questId, out QuestModel quest))
+            {
+                Debug.LogError($"Couldn't find quest with ID {questId} in DataStore");
+                return;
+            }
+
+            if(!questEntries.TryGetValue(questId, out QuestsPanelEntry questEntry))
+            {
+                questEntry = Instantiate(questPrefab, questsContainer).GetComponent<QuestsPanelEntry>();
+                questEntry.OnReadMoreClicked += ShowQuestPopup;
+                questEntries.Add(questId, questEntry);
+            }
+
+            questEntry.Populate(quest);
+            layoutRebuildRequested = true;
+        }
+
+        public void RemoveQuest(string questId)
+        {
+            questsToBeAdded.Remove(questId);
+
+            if(!questEntries.TryGetValue(questId, out QuestsPanelEntry questEntry))
+                return;
+
+            questEntries.Remove(questId);
+            Destroy(questEntry.gameObject);
+
+            if(currentQuestInPopup == questId)
+                questPopup.Close();
+        }
+
+        public void ClearQuests()
+        {
+            questPopup.Close();
+            foreach (QuestsPanelEntry questEntry in questEntries.Values)
+            {
+                Destroy(questEntry.gameObject);
+            }
+            questEntries.Clear();
+            questsToBeAdded.Clear();
+        }
+
+        internal void ShowQuestPopup(string questId)
+        {
+            if (!quests.TryGetValue(questId, out QuestModel quest))
+            {
+                Debug.Log($"Couldnt find quest with id {questId}");
+                return;
+            }
+
+            currentQuestInPopup = questId;
+            questPopup.Populate(quest);
+            questPopup.Show();
+        }
+
+        internal void Update()
+        {
+            if (layoutRebuildRequested)
+            {
+                layoutRebuildRequested = false;
+                Utils.ForceRebuildLayoutImmediate(questsContainer);
+            }
+
+            for (int i = 0; i < ENTRIES_PER_FRAME && questsToBeAdded.Count > 0; i++)
+            {
+                string questId = questsToBeAdded.First();
+                questsToBeAdded.RemoveAt(0);
+                AddOrUpdateQuest(questId);
+            }
```

Comment:
This kind of data management logic shouldn't be in a controller?

Reply:
I have been having this argument with myself multiple times hahaha. 

The reason we are doing this is because we are splitting the instantiation of quest entries by frames. I ended up thinking that's the view's responsibility and not the controller. IMO the controller should tell the view "these are the entries you have to show" and the view should deal with them, in this case processing them asynchronously.